src/environmentd/src/http/metrics.rs

```diff
@@ -0,0 +1,229 @@
+}
+
+impl<S, R> Service<R> for PrometheusService<S>
+where
+    S: Service<R>,
+    S::Response: IntoResponse,
+    S::Error: Into<Infallible>,
+    S::Future: Send,
+{
+    type Error = S::Error;
+    type Response = axum::response::Response;
+    type Future = PrometheusFuture<S::Future>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: R) -> Self::Future {
+        let fut = self.service.call(req);
+        PrometheusFuture::new(fut, self.path, self.metrics.clone())
+    }
+}
+
+#[pin_project]
+struct PrometheusFuture<F> {
+    /// The name of the endpoint that this request corresponds to.
+    path: &'static str,
+    /// Instant at which we started the requst.
+    timer: Option<HistogramTimer>,
+    /// Metrics registry used to record events.
+    metrics: Metrics,
+    /// Inner request future.
+    #[pin]
+    fut: F,
+}
+
+impl<F> PrometheusFuture<F> {
+    fn new(fut: F, path: &'static str, metrics: Metrics) -> Self {
+        PrometheusFuture {
+            path,
+            timer: None,
+            metrics,
+            fut,
+        }
+    }
+}
+
+impl<F, R, E> Future for PrometheusFuture<F>
+where
+    R: IntoResponse,
+    E: Into<Infallible>,
+    F: Future<Output = Result<R, E>>,
+{
+    type Output = Result<axum::response::Response, E>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        if this.timer.is_none() {
+            // Start timer so we can track duration of request.
+            let duration_metric = this
+                .metrics
+                .request_duration
+                .with_label_values(&[this.path]);
+            *this.timer = Some(duration_metric.start_timer());
+
+            // Increment our counter of currently active requests.
+            this.metrics
+                .requests_active
+                .with_label_values(&[this.path])
+                .inc();
+        }
+
+        // Poll the inner future to make progress.
+        match this.fut.poll(cx) {
+            Poll::Ready(resp) => {
+                let ok = resp.infallible_unwrap();
+                let resp = ok.into_response();
+                let status = resp.status();
+
+                // Record the completion of this request.
+                this.metrics
+                    .requests
+                    .with_label_values(&[this.path, status.as_str()])
+                    .inc();
+
+                // Record the duration of this request.
+                if let Some(timer) = this.timer.take() {
+                    timer.observe_duration();
+                }
+
+                // We've completed this request, so decrement the count.
+                this.metrics
+                    .requests_active
+                    .with_label_values(&[this.path])
+                    .dec();
```

Comment:
you'll also technically need to `dec` this in a `Drop` impl, if we weren't polled to completion

Reply:
Good call! I added a Drop impl and a test to assert the right behavior